ext/drx/scatter_gather_aarch64.c

```diff
@@ -673,11 +709,50 @@ bool
 drx_scatter_gather_restore_state(void *drcontext, dr_restore_state_info_t *info,
                                  instr_t *sg_inst)
 {
-    DR_ASSERT(instr_is_gather(sg_inst) || instr_is_scatter(sg_inst));
-    /* TODO i#5365, i#5036: Restore the scratch predicate register.
+    /* If this function is called, we know that a fault occurred on an instruction in a
+     * fragment which expands a scatter/gather instruction, but we don't know whether
+     * the instruction that faulted was one of the expansion loads or stores emitted by
+     * drx_expand_scatter_gather(), or part of instrumentation added later by a client.
+     *
+     * If a scatter/gather expansion instruction faults we need to treat it as if the
+     * expanded scatter/gather instruction had faulted and set the register state as
+     * appropriate for the expanded instruction. This isn't implemented yet so we hit
+     * an assert below.
+     *
+     * Previously this function would always assert but this causes a problem with
+     * clients (such as memval_simple) that use drx_buf (or similar) which uses faulting
+     * stores to manage the trace buffer.
+     * Until we implement proper state restoration we need to filter out faults that
+     * don't come from scatter/gather expansion instructions and pass them on to the
+     * client to handle, otherwise we can get spurious failures with clients like
+     * memval_simple.
+     */
+    if (info->fragment_info.ilist != NULL) {
+        byte *pc = info->fragment_info.cache_start_pc;
+        for (instr_t *instr = instrlist_first(info->fragment_info.ilist); instr != NULL;
+             instr = instr_get_next(instr)) {
+            if (pc == info->raw_mcontext->pc && !instr_is_label(instr)) {
+                /* Found the faulting instruction */
+                if (!scatter_gather_is_expanded_ld_st(instr)) {
```

Comment:
I wonder if we can use drmgr_is_emulation_start and drmgr_is_emulation_end instead for this. Hmm.. but that would not help us differentiate between drmemtrace-added instrumentation and the expansion itself I guess.

Reply:
As you say, that wouldn't let us distinguish the expansion instructions from instructions added by later instrumentation stages.